nixos/modules/security/duosec.nix

```diff
@@ -194,22 +183,59 @@ in
     };
   };
 
-  config = mkIf (cfg.ssh.enable || cfg.pam.enable) {
-     environment.systemPackages = [ pkgs.duo-unix ];
-
-     security.wrappers.login_duo.source = "${pkgs.duo-unix.out}/bin/login_duo";
-     environment.etc = loginCfgFile // pamCfgFile;
-
-     /* If PAM *and* SSH are enabled, then don't do anything special.
-     If PAM isn't used, set the default SSH-only options. */
-     services.openssh.extraConfig = mkIf (cfg.ssh.enable || cfg.pam.enable) (
-     if cfg.pam.enable then "UseDNS no" else ''
-       # Duo Security configuration
-       ForceCommand ${config.security.wrapperDir}/login_duo
-       PermitTunnel no
-       ${optionalString (!cfg.allowTcpForwarding) ''
-         AllowTcpForwarding no
-       ''}
-     '');
-  };
+  config = mkMerge [
+    {
+      system.activationScripts = {
+        login_duo = if cfg.ssh.enable then ''
+          mkdir -m 0755 -p /etc/duo
+
+          umask 277
+          conf="$(mktemp)"
+          {
+            cat ${pkgs.writeText "login_duo.conf" configFileLogin}
+            printf 'skey = %s\n' "$(cat ${cfg.secretKeyFile})"
+          } >"$conf"
+
+          chown sshd "$conf"
+          mv -fT "$conf" /etc/duo/login_duo.conf
+        '' else ''
+          rm -f /etc/duo/login_duo.conf
```

Comment:
Hmm that has me slightly worried now, because what if there are people who didn't use the NixOS module to manage this file

Reply:
Yes something I considered but I'm not sure how to get around that if using `/etc/duo`.

Another potential alternative is to store the file elsewhere. From upstream documentation it *looks* like the config file location can be specified. I'm not sure where a great spot would be though. In the case of `login_duo.conf` this is only used by `ssh.service` so I could piggyback of the `ssh.service` and add `RuntimeDirectory = duo;` then a `preStart` script. It starts to feel a bit imposing on the `ssh` module, but probably worth it in this case.

I can't stress enough how bad it is that `skey` is in the nix store. This **has** to be fixed, so pick the solution that causes the least turbulence?